src/init_parsing.jl

```diff
@@ -3,147 +3,185 @@ struct HeaderParsingError <: Exception
 end
 Base.showerror(io::IO, e::HeaderParsingError) = print(io, e.msg)
 
-function initial_read_and_lex_and_skip!(io, chunking_ctx, settings, escapechar, openquotechar, closequotechar, ignoreemptyrows)
+# We might need to skip to get to the header and then skip again to get to the data
+# This function does the *first* skip and initializes the lexer
+function initial_read_and_lex_and_skip!(io, chunking_ctx, input_args, escapechar, openquotechar, closequotechar, ignoreemptyrows)
     # First ingestion of raw bytes from io
     bytes_read_in = ChunkedBase.initial_read!(io, chunking_ctx)
 
     # We need to detect the newline first to construct the Lexer
-    newline = settings.newlinechar
+    newline = input_args.newlinechar
     if isnothing(newline)
         newline = ChunkedBase._detect_newline(chunking_ctx.bytes, 1, bytes_read_in)
     end
 
-    lexer = settings.no_quoted_newlines ?
+    lexer = input_args.no_quoted_newlines ?
         Lexer(io, nothing, newline) :
         Lexer(io, escapechar, openquotechar, closequotechar, newline)
 
     # Find newlines
     ChunkedBase.initial_lex!(lexer, chunking_ctx, bytes_read_in)
 
     # First skip over commented lines, then jump to header / data row
-    should_parse_header = settings.header_at > 0
-    pre_header_skiprows = max(0, (should_parse_header ? settings.header_at : settings.data_at) - 1)
+    should_parse_header = input_args.header_at > 0
+    pre_header_skiprows = max(0, (should_parse_header ? input_args.header_at : input_args.data_at) - 1)
     lines_skipped_total = ChunkedBase.skip_rows_init!(lexer, chunking_ctx, pre_header_skiprows, ignoreemptyrows)
 
     return lexer, lines_skipped_total
 end
 
+# This function does the *second* skip and initializes the parsing context.
+# The parsing context requires us knowing the schema and the header, so we
+# need to parse the header first. We don't do type inference yet, so any
+# unknown types will be filled with String.
 function process_header_and_schema_and_finish_row_skip!(
     parsing_ctx::ParsingContext,
     chunking_ctx::ChunkingContext,
     lexer::Lexer,
     schema_input,
-    settings::ParserSettings,
+    input_args::InputArguments,
     lines_skipped_total::Int
 )
     input_is_empty = length(chunking_ctx.newline_positions) == 1
     options = parsing_ctx.options
     ignorerepeated = options.ignorerepeated
 
-    header_provided = !isnothing(settings.header)
+    header_provided = !isnothing(input_args.header)
     schema_provided = schema_input isa Vector{DataType}
-    should_parse_header = settings.header_at > 0
+    should_parse_header = input_args.header_at > 0
     schema = parsing_ctx.schema
     schema_provided && validate_schema(schema_input)
+    newlines = chunking_ctx.newline_positions
 
     @inbounds if schema_provided & header_provided
```

Comment:
this is a _big_ block of code to assert is all `@inbounds`

have we definitely verified (a) that's guaranteed (e.g. there no way user inputs could affect that?), and (b) it matters for performance?